packages/kurt-vertex-ai/src/KurtVertexAI.ts

```diff
@@ -265,31 +265,60 @@ async function* transformStreamWithOptionalTools<
     for (const [partIndex, part] of parts.entries()) {
       const chunk = part.text
       const isFinal =
-        (isContentFinal && partIndex === parts.length - 1) || part.functionCall
+        (isContentFinal || part.functionCall) && partIndex === parts.length - 1
 
       if (chunk) {
         chunks.push(chunk)
         yield { chunk }
       }
       if (isFinal) {
-        const { functionCall } = part
-        if (functionCall) {
-          const { name } = functionCall
-          const schema = tools[name]
-          if (!schema) {
-            throw new Error(
-              `Vertex AI tried to call tool ${name} which isn't in the tool set ${JSON.stringify(
-                Object.keys(tools)
-              )}}`
-            )
+        if (part.functionCall) {
+          const allData = parts.map((part) => {
+            if (!part.functionCall) {
+              throw new Error(
+                `Vertex AI mixed function calls with non-function calls in the same raw stream event: ${JSON.stringify(
+                  rawEvent
+                )}`
+              )
+            }
+
+            const { name } = part.functionCall
+
+            const schema = tools[name]
+            if (!schema) {
+              throw new Error(
+                `Vertex AI tried to call tool ${name} which isn't in the tool set ${JSON.stringify(
+                  Object.keys(tools)
+                )}}`
+              )
+            }
+            return {
+              name,
+              args: applySchemaToFuzzyStructure(schema, part.functionCall),
+            } as D
+          })
+
+          // Emit a text chunk for each tool call (with line breaks in between).
+          for (const [dataIndex, data] of allData.entries()) {
+            if (dataIndex > 0) {
+              chunks.push("\n")
+              yield { chunk: "\n" }
+            }
+            const text = JSON.stringify(data.args)
+            chunks.push(text)
+            yield { chunk: text }
+          }
+
+          // biome-ignore lint/style/noNonNullAssertion: we already validated above that length > 0
+          const data = allData[0]!
+          const additionalData = allData.slice(1)
```

Comment:
Don't have to but you could do `const [data, ...additionalData] = allData;`

Reply:
Would probably need to do something like:
```
const [maybeData, ...additionalData] = allData
// biome-ignore lint/style/noNonNullAssertion: we already validated above that length > 0
const data = maybeData!
```

Does that look better or worse than the current one?

Reply:
Check my comment above for a way to deal with the assertion, but yes it would be like this, I like the destructuring syntax better, but no strong opinions on this one.